rust_src/src/dispnew.rs

```diff
@@ -105,4 +105,20 @@ pub extern "C" fn set_window_update_flags(w: LispWindowRef, on_p: bool) {
     }
 }
 
+/***********************************************************************
+			   Blinking cursor
+ ***********************************************************************/
+
+#[lisp_fn]
+pub fn internal_show_cursor(window: LispObject, show: LispObject) {
+    if !unsafe { redisplaying_p } {
+        window_or_selected(window).set_cursor_off_p(show.is_nil())
+    }
```

Comment:
This function returns `Qnil`.

Reply:
Making no return is the equivalent. I prefer the explicit return too.

Reply:
Right, we discussed this recently.

Reply:
What's the point of the explicit return if it's always returning `Qnil` so it can just return something? That reasoning is why it's elided with `()` in Rust—something has to be returned, but it's just noise (unless it's a sum arm, but you'd use `LispObject` for that).

Reply:
Clarity. I am not a fan of magic in my code personally. I'd be ok with the ()/Qnil case being dropped from `lisp_fn`.

If our mission is having a project open to new developers and maybe casual coders, the more obvious our code is the better off we will be.

Reply:
If we're trying to be anti-magic here, shouldn't we be using `-> ()` on all Rust functions, then? I don't see how a mental continuation of Rust functions that always return values into Lisp functions is broken here. If someone working on Remacs doesn't know that "void" Rust functions actually return unit, I think it's more valuable to tell them that in the porting guide while noting how "void" Lisp functions are declared and written than to clutter up the code with trailing `Qnil`s, like C has to do, and Rust almost had. This way, the existing pragmatic clarity of Rust's implicit unit return is kept while Rust code they may run into in the future that deals with the output of "void" functions is made less surprising. Additionally, this means that explicit `Qnil` returns are more likely to denote the termination of a list, and not just void, better conveying intent with Rust's type system.

Reply:
@bb010g I think you're right. `lisp_fn` already contains a lot of magic and newcomers already have to take a look at the wiki. But we are using `-> !` though I don't know why it's not `-> ()`. Doesn't really matter I guess.

Reply:
`-> !` means "function never returns to caller" not "does not return a value". Anything that returns '!' is an executional dead end and the compiler can use this information.

`-> ()` means 'does not return a value'.

Reply:
Not my day. `-> ()` seems good.